module/apmfasthttp/server_test.go

```diff
@@ -30,67 +32,159 @@ import (
 	"github.com/valyala/fasthttp"
 
 	"go.elastic.co/apm/module/apmfasthttp/v2"
+	"go.elastic.co/apm/v2/model"
 	"go.elastic.co/apm/v2/transport/transporttest"
 )
 
-func TestServerHTTPResponse(t *testing.T) {
+type assertFunc func(model.Transaction, *http.Response)
+
+func testServer(t *testing.T, s *fasthttp.Server, wg *sync.WaitGroup, assertFn assertFunc) {
+	t.Helper()
+
 	tracer, transport := transporttest.NewRecorderTracer()
 	defer tracer.Close()
 
-	handler := func(ctx *fasthttp.RequestCtx) {
-		ctx.Error(fasthttp.StatusMessage(fasthttp.StatusUnauthorized), fasthttp.StatusUnauthorized)
-	}
-	handler = apmfasthttp.Wrap(handler, apmfasthttp.WithTracer(tracer))
 	shutdown := make(chan error)
 	defer close(shutdown)
+
 	ln, err := net.Listen("tcp", "127.0.0.1:")
 	require.NoError(t, err)
 
-	s := &fasthttp.Server{
-		Handler: handler,
-		Name:    "test-server",
-	}
+	s.Handler = apmfasthttp.Wrap(s.Handler, apmfasthttp.WithTracer(tracer))
 
 	go func() {
 		shutdown <- s.Serve(ln)
 	}()
 
 	addr := ln.Addr().String()
+
 	resp, err := http.Get("http://" + addr)
 	require.NoError(t, err)
-	io.Copy(ioutil.Discard, resp.Body) // consume body to complete request handler
+
+	_, _ = io.Copy(ioutil.Discard, resp.Body) // consume body to complete request handler
 	resp.Body.Close()
-	assert.Equal(t, fasthttp.StatusUnauthorized, resp.StatusCode)
+
+	if wg != nil {
+		wg.Wait()
+	}
+
 	tracer.Flush(nil)
 	payloads := transport.Payloads()
 
-	// the transaction is ended after the response body is fully written,
-	// so a single call to `tracer.Flush()` may not have an event yet
-	// enqueued. Continue to call `tracer.Flush()` and wait for the
-	// transaction to have ended.
-	// This is unique to fasthttp's implementation.
-	timer := time.NewTimer(100 * time.Millisecond)
-	defer timer.Stop()
-	for {
-		select {
-		case <-timer.C:
-			t.Fatal("timed out waiting for payload")
-		default:
+	count := 0
+	for len(payloads.Transactions) == 0 {
+		// the transaction is ended after the response body is fully written,
+		// so a single call to `tracer.Flush()` may not have an event yet
+		// enqueued. Continue to call `tracer.Flush()` and wait for the
+		// transaction to have ended.
+		// This is unique to fasthttp's implementation.
+		if count > 10 {
+			t.Fatal("no transactions found")
 		}
-		if len(payloads.Transactions) > 0 {
-			break
-		}
-		time.Sleep(10 * time.Millisecond)
+
+		count++
+
+		time.Sleep(100 * time.Millisecond)
 		tracer.Flush(nil)
 		payloads = transport.Payloads()
 	}
```

Comment:
Why has this changed? Doesn't seem related to the PR. I'm concerned that your changes may lead to failures on slow CI machines, so please revert this unless it's needed.

Reply:
Done!